### PR TITLE
Nikita branch

### PR DESCRIPTION
**Pull request by Nikita3A**
Added authentication component and auth Middleware to user component

### CODE REVIEW

**Comment by Nikita3A** [comment]
Added authentication component and auth Middleware

File: .eslintrc.js [hunk 1]
```diff
@@ -12,5 +12,9 @@ module.exports = {
     },
     rules: {
         indent: ['error', 4],
+        'no-unused-vars': 'warn',
```
Comment (v-checha) [inline .eslintrc.js]: thats vary important rules, you can't disable it

File: package.json [hunk 2]
```diff
@@ -18,11 +18,13 @@
   "homepage": "https://github.com/ChechaValerii/intership-template#readme",
   "dependencies": {
     "@hapi/joi": "^17.1.1",
+    "bcrypt": "^5.0.1",
```
Comment (v-checha) [inline package.json]: remove corner

File: package.json [hunk 3]
```diff
@@ -18,11 +18,13 @@
...
     "compression": "1.7.4",
     "cookie-parser": "1.4.6",
     "cors": "2.8.5",
     "express": "4.17.1",
     "helmet": "4.6.0",
+    "jsonwebtoken": "^8.5.1",
```
Comment (v-checha) [inline package.json]: the same

File: src/components/User/authMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// module.exports = function verifyToken(req, res, next) {
```
Comment (v-checha) [inline src/components/User/authMiddleware.js]: remove comments

File: src/components/User/authMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+//         next();
+//     } else {
+//         // Forbidden
+//         res.sendStatus(403);
+//     }
+// };
+module.exports.verifyToken = function (req, res, next) {
+    // Get auth header value
+    // const bearerHeader = req.headers['authorization'];
+    const bearerHeader = req.headers.authorization;
```
Comment (v-checha) [inline src/components/User/authMiddleware.js]: please try use `req.get(field)`

File: src/components/User/index.js [hunk 6]
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable import/no-unresolved */
```
Comment (v-checha) [inline src/components/User/index.js]: remove this comment

File: src/components/User/index.js [hunk 7]
```diff
@@ -11,12 +14,19 @@ const ValidationError = require('../../error/ValidationError');
  */
 async function findAll(req, res, next) {
     try {
-        const users = await UserService.findAll();
-
-        res.status(200).json({
-            data: users,
-        });
+        const verification = jwt.verify(req.token, 'secretkey');
```
Comment (v-checha) [inline src/components/User/index.js:15]: move to validation middleware like this
`app.get('/books', authenticateJWT, (req, res) => {
    res.json(books);
});`

File: src/components/User/index.js [hunk 8]
```diff
@@ -11,12 +14,19 @@ const ValidationError = require('../../error/ValidationError');
...
+        if (verification) {
+            const users = await UserService.findAll();
+            res.status(200).json({
+                data: users,
+            });
+        } else {
+            // console.log(err);
```
Comment (v-checha) [inline src/components/User/index.js]: remove log

File: src/components/User/index.js [hunk 9]
```diff
@@ -11,12 +14,19 @@ const ValidationError = require('../../error/ValidationError');
...
+            console.log('Forbidden to watch users');
+            res.sendStatus(403);
```
Comment (v-checha) [inline src/components/User/index.js]: set status + send message
all client responses must be in JSON formats

File: src/components/User/index.js [hunk 10]
```diff
@@ -11,12 +14,19 @@ const ValidationError = require('../../error/ValidationError');
...
+        }
     } catch (error) {
+        console.log('Error in find all');
```
Comment (v-checha) [inline src/components/User/index.js]: remove console.log

File: src/components/User/index.js [hunk 11]
```diff
@@ -35,17 +45,22 @@ async function findAll(req, res, next) {
  */
 async function findById(req, res, next) {
     try {
-        const { error } = UserValidation.findById(req.params);
+        const verification = jwt.verify(req.token, 'secretkey');
```
Comment (v-checha) [inline src/components/User/index.js:44]: the same

File: src/components/User/index.js [hunk 12]
```diff
@@ -35,17 +45,22 @@ async function findAll(req, res, next) {
...
+            const user = await UserService.findById(req.params.id);
 
-        return res.status(200).json({
-            data: user,
-        });
+            return res.status(200).json({
+                data: user,
+            });
+        }
+        console.log('Forbidden to update user');
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/User/index.js [hunk 13]
```diff
@@ -72,17 +87,22 @@ async function findById(req, res, next) {
  */
 async function create(req, res, next) {
     try {
-        const { error } = UserValidation.create(req.body);
+        const verification = jwt.verify(req.token, 'secretkey');
+        if (verification) {
```
Comment (v-checha) [inline src/components/User/index.js:80]: the same

File: src/components/User/index.js [hunk 14]
```diff
@@ -72,17 +87,22 @@ async function findById(req, res, next) {
...
+            const user = await UserService.create(req.body);
 
-        return res.status(200).json({
-            data: user,
-        });
+            return res.status(200).json({
+                data: user,
+            });
+        }
+        console.log('Forbidden to create user');
```
Comment (v-checha) [inline src/components/User/index.js]: the same

File: src/components/User/index.js [hunk 15]
```diff
@@ -109,17 +129,22 @@ async function create(req, res, next) {
  */
 async function updateById(req, res, next) {
     try {
-        const { error } = UserValidation.updateById(req.body);
+        const verification = jwt.verify(req.token, 'secretkey');
```
Comment (v-checha) [inline src/components/User/index.js:114]: the same

File: src/components/User/index.js [hunk 16]
```diff
@@ -109,17 +129,22 @@ async function create(req, res, next) {
...
+            }
 
-        const updatedUser = await UserService.updateById(req.body.id, req.body);
+            const updatedUser = await UserService.updateById(req.body.id, req.body);
 
-        return res.status(200).json({
-            data: updatedUser,
-        });
+            return res.status(200).json({
+                data: updatedUser,
```
Comment (v-checha) [inline src/components/User/index.js:119]: the same

File: src/components/User/index.js [hunk 17]
```diff
@@ -146,17 +171,64 @@ async function updateById(req, res, next) {
  */
 async function deleteById(req, res, next) {
     try {
-        const { error } = UserValidation.deleteById(req.body);
+        const verification = jwt.verify(req.token, 'secretkey');
```
Comment (v-checha) [inline src/components/User/index.js:149]: the same

File: src/components/User/index.js [hunk 18]
```diff
@@ -146,17 +171,64 @@ async function updateById(req, res, next) {
...
+
+        res.status(500).json({
+            message: error.name,
+            details: error.message,
+        });
+
+        return next(error);
+    }
+}
+// My code ////////////////////////
```
Comment (v-checha) [inline src/components/User/index.js]: remove this coment

File: src/components/User/index.js [hunk 19]
```diff
@@ -146,17 +171,64 @@ async function updateById(req, res, next) {
...
 
         if (error) {
             throw new ValidationError(error.details);
         }
 
-        const deletedUser = await UserService.deleteById(req.body.id);
-
-        return res.status(200).json({
-            data: deletedUser,
+        // const loggedUser = await UserService.deleteById(req.body.id);
```
Comment (v-checha) [inline src/components/User/index.js]: remove comment

File: src/components/User/index.js [hunk 20]
```diff
@@ -146,17 +171,64 @@ async function updateById(req, res, next) {
...
+        const userLoginData = req.body;
+        console.log(userLoginData);
```
Comment (v-checha) [inline src/components/User/index.js]: remove log

File: src/components/User/index.js [hunk 21]
```diff
@@ -146,17 +171,64 @@ async function updateById(req, res, next) {
...
+        jwt.sign({ userLoginData }, 'secretkey', { expiresIn: '1m' }, (_error, token) => {
+            res.json({ token });
         });
+        // return res.send('Logging user');
```
Comment (v-checha) [inline src/components/User/index.js]: remove log

File: src/components/User/index.js [hunk 22]
```diff
@@ -173,11 +245,15 @@ async function deleteById(req, res, next) {
...
 }
+// My code ////////////////////////
 
 module.exports = {
     findAll,
     findById,
     create,
     updateById,
     deleteById,
+    // My code //
```
Comment (v-checha) [inline src/components/User/index.js]: remove comment

File: src/components/User/index.js [hunk 23]
```diff
@@ -173,11 +245,15 @@ async function deleteById(req, res, next) {
...
+    loginUser,
+    // My code //
```
Comment (v-checha) [inline src/components/User/index.js]: remove comment

File: src/components/User/model.js [hunk 24]
```diff
@@ -1,6 +1,12 @@
+/* eslint-disable consistent-return */
+/* eslint-disable func-names */
```
Comment (v-checha) [inline src/components/User/model.js]: remove comment

File: src/components/User/model.js [hunk 25]
```diff
@@ -1,6 +1,12 @@
...
 const { Schema } = require('mongoose');
+// eslint-disable-next-line import/no-unresolved
```
Comment (v-checha) [inline src/components/User/model.js]: remove comment

File: src/components/auth/index.js [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+            throw new ValidationError(error.details);
+        }
+
+        const user = await AuthService.signUp(req.body);
+
+        return res.status(200).json({
+            data: user,
+        });
+    } catch (error) {
+        console.log('Not sign up!');
```
Comment (v-checha) [inline src/components/auth/index.js]: remove comment

File: src/components/auth/index.js [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+
+        if (validPassword) {
+            console.log('Pass is correct');
+        } else {
+            console.log('Wroing password in auth/index.js');
+            res.send('Wrong password');
+        }
+
+        console.log({ email, password });
+        // jwt.sign({ email, password }, 'secretkey', { expiresIn: '1m' }, (_error, token) => {
```
Comment (v-checha) [inline src/components/auth/index.js]: remove comment

File: src/components/auth/index.js [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+        const { email, password } = req.body;
+        const user = await userModel.findOne({ email });
+        const validPassword = bcrypt.compareSync(password, user.password);
+
+        if (validPassword) {
+            console.log('Pass is correct');
+        } else {
+            console.log('Wroing password in auth/index.js');
```
Comment (v-checha) [inline src/components/auth/index.js]: remove log

File: src/components/auth/index.js [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+            res.send('Wrong password');
+        }
+
+        console.log({ email, password });
```
Comment (v-checha) [inline src/components/auth/index.js]: remove log

File: src/components/auth/index.js [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+        // jwt.sign({ email, password }, 'secretkey', { expiresIn: '1m' }, (_error, token) => {
+        //     res.json({ token });
+        // });
+        const accessToken = jwt.sign({ email, password }, 'secretkey', { expiresIn: '1m' });
+        const refreshToken = jwt.sign({ email, password }, 'RefreshSecretkey', { expiresIn: '10h' });
+        refreshTokens.push(refreshToken);
```
Comment (v-checha) [inline src/components/auth/index.js:72]: need save to database

File: src/components/auth/index.js [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+/**
+ * @function
+ * @param {express.Request} req
+ * @param {express.Response} res
+ * @param {express.NextFunction} next
+ * @returns {Promise<void>}
+ */
+async function refreshToken(req, res, next) {
+    try {
+        const refToken = req.body.refreshToken;
```
Comment (v-checha) [inline src/components/auth/index.js:101]: please use Joi validation

File: src/components/User/index.js (hunk 6 above)
Comment (Nikita3A) [reply src/components/User/index.js]: Done

**Pull request closed without merging**